OpenRA.Game/Traits/TraitsInterfaces.cs

```diff
@@ -170,8 +170,31 @@ public interface IVoiced
 		bool HasVoice(Actor self, string voice);
 	}
 
-	[RequireExplicitImplementation]
-	public interface IStoreResources { int Capacity { get; } }
+	public interface IStoresResourcesInfo : ITraitInfoInterface
+	{
+		string[] ResourceTypes { get; }
+	}
+
+	public interface IStoresResources
+	{
+		bool HasType(string resourceType);
+
+		/// <summary>The amount of resources that can be stored.</summary>
+		int Capacity { get; }
+
+		/// <summary>Stored resources.</summary>
+		/// <remarks>Dictionary key refers to resourceType, value refers to resource amount.</remarks>
+		IReadOnlyDictionary<string, int> Contents { get; }
```

Comment:
I don't think the extra complexity is worth it. (At least unless we have some numbers on that this makes a significant impact.)

Reply:
It's not a cold path, but also it's not very hot, so such optimisations shouldn't have very noticeable impact. It may make sense to do an optimisation pass after the whole refactor is finished, but atm it would just make the code much harder to read.

Reply:
8 haversters times 8 players times any resource update.